File: backend/app/schemas/leave.py
from pydantic import BaseModel, Field
from typing import Optional, List
from datetime import datetime


LEAVE_TYPES = ["Annual", "Sick", "Casual", "Maternity", "Paternity", "Emergency", "Unpaid", "Compensatory", "Other"]
CALCULATIONS = ["MONTHLY", "YEARLY", "QUARTERLY"]


class LeaveMasterCreate(BaseModel):
    leave_type:    str = Field(..., min_length=1, max_length=100)
    entitled:      int = Field(..., ge=0)
    calendar_year: int = Field(..., ge=2000, le=2100)
    calculation:   str = Field(..., min_length=1, max_length=50)


class LeaveMasterUpdate(BaseModel):
    leave_type:    Optional[str] = Field(None, min_length=1, max_length=100)
    entitled:      Optional[int] = Field(None, ge=0)
    calendar_year: Optional[int] = Field(None, ge=2000, le=2100)
    calculation:   Optional[str] = Field(None, min_length=1, max_length=50)


class LeaveMasterOut(BaseModel):
    leave_master_id: int
    tenant_id:       int
    leave_type:      str
    entitled:        int
    calendar_year:   int
    calculation:     str
    created_at:      datetime
    updated_at:      datetime

    model_config = {"from_attributes": True}


class PaginatedLeaveMasters(BaseModel):
    items:       List[LeaveMasterOut]
    total:       int
    page:        int
    page_size:   int
    total_pages: int

File: frontend/src/api/leave.ts
import api from './client'

export interface LeaveMaster {
  leave_master_id: number
  tenant_id:       number
  leave_type:      string
  entitled:        number
  calendar_year:   number
  calculation:     string
  created_at:      string
  updated_at:      string
}

export interface LeaveMasterPayload {
  leave_type:    string
  entitled:      number
  calendar_year: number
  calculation:   string
}

export interface PaginatedLeaveMasters {
  items:       LeaveMaster[]
  total:       number
  page:        number
  page_size:   number
  total_pages: number
}

export const LEAVE_TYPES = [
  'Annual', 'Sick', 'Casual', 'Maternity', 'Paternity',
  'Emergency', 'Unpaid', 'Compensatory', 'Other',
]

export const CALCULATIONS = ['MONTHLY', 'YEARLY', 'QUARTERLY']

export const leaveMasterApi = {
  list: (params?: { page?: number; page_size?: number; calendar_year?: number }) =>
    api.get<PaginatedLeaveMasters>('/leave-master', { params }).then(r => r.data),

  create: (data: LeaveMasterPayload) =>
    api.post<LeaveMaster>('/leave-master', data).then(r => r.data),

  update: (id: number, data: Partial<LeaveMasterPayload>) =>
    api.patch<LeaveMaster>(`/leave-master/${id}`, data).then(r => r.data),

  delete: (id: number) =>
    api.delete(`/leave-master/${id}`).then(r => r.data),
}
